File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/MediaAssetRecord.java
package in.org.quicko.silo.client.media;

import java.time.Instant;
import java.util.List;

/**
 * One catalogued asset's mapped state — {@link MediaMapper}'s output and the
 * value {@link MediaAsset} holds.
 *
 * <p>{@code reference} is what belongs in an entry field; {@code url} is only
 * where to fetch today's bytes, and a rename or a move can change it.
 */
public record MediaAssetRecord(
    String id,
    String filename,
    String folder,
    String blobKey,
    long sizeInBytes,
    String contentType,
    String hash,
    MediaState state,
    List<String> tags,
    String url,
    long usageCount,
    Instant createdAt,
    Instant updatedAt) {

  public MediaAssetRecord {
    tags = tags == null ? List.of() : List.copyOf(tags);
  }
}

File: packages/silo-client-dart/lib/src/media/media_asset_record.dart
import '../transport/json_values.dart';
import '../transport/timestamps.dart';
import 'media_state.dart';

/// Everything an asset is, as a value.
final class MediaAssetRecord {
  const MediaAssetRecord({
    required this.id,
    required this.filename,
    required this.folder,
    required this.blobKey,
    required this.sizeInBytes,
    required this.contentType,
    required this.hash,
    required this.state,
    required this.tags,
    required this.url,
    required this.usageCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MediaAssetRecord.fromJson(Map<String, Object?> json) => MediaAssetRecord(
    id: json['id'] as String,
    filename: json['filename'] as String,
    folder: json['folder'] as String,
    blobKey: json['blob_key'] as String,
    sizeInBytes: json['size'] as int,
    contentType: json['content_type'] as String,
    hash: json['hash'] as String,
    state: MediaState.of(json['state']),
    tags: JsonValues.strings(json['tags']),
    url: json['url'] as String,
    usageCount: json['usage_count'] as int? ?? 0,
    createdAt: Timestamps.parse(json['created_at']),
    updatedAt: Timestamps.parse(json['updated_at']),
  );

  final String id;
  final String filename;
  final String folder;
  final String blobKey;
  final int sizeInBytes;
  final String contentType;
  final String hash;
  final MediaState state;
  final List<String> tags;
  final String url;
  final int usageCount;
  final DateTime createdAt;
  final DateTime updatedAt;
}
